src/asynchronous/server.rs

```diff
@@ -381,12 +381,14 @@ impl ReaderDelegate for ServerReader {
     async fn handle_msg(&self, msg: GenMessage) {
         let handler_shutdown_waiter = self.handler_shutdown.subscribe();
         let context = self.context();
+        let (wait_tx, wait_rx) = tokio::sync::oneshot::channel::<()>();
```

Comment:
More make sence to use [tokio::sync::Barrier](https://docs.rs/tokio/latest/tokio/sync/struct.Barrier.html)?

Reply:
Barrier is for blocking a number of tasks, but here we just need to block the current task, it seems the oneshot channel is a simpler choice, I think.

Reply:
Maybe we can use tokio::sync::Notify, but as it do not implement Drop so that we have to call `notify_one()` in a lot of places otherwise the `handle_msg` method of `ServerReader` will be blocked if we miss calling `notify_one()` in any code branches, hence I think it is easy to introduce bugs. 
I think the oneshot::channel is a better choice because whenever it is droped, the wait of it will be unblocked. so we don't have to call drop everywhere.

Reply:
oneshot is ok in my opinion